r2dbc-spi/src/main/java/io/r2dbc/spi/Result.java

```diff
@@ -49,4 +54,124 @@
      */
     <T> Publisher<T> map(BiFunction<Row, RowMetadata, ? extends T> mappingFunction);
 
+    /**
+     * Returns a filtered variant of the {@link Result} to selectively consume result segments matching {@link Predicate filter predicate}.
+     * <p>The returned {@link Result} is a potentially reduced view of the underlying {@link Result} to filter out unwanted result segments.  For example, filtering {@link Message} segments of
+     * {@link Message.Severity#ERROR error severity} from the result
+     * lets the result complete without an error.
+     *
+     * @param filter the non-interfering and stateless {@link Predicate} to apply to each element to determine if it should be included
+     * @return a filtered {@link Result}
+     * @throws IllegalArgumentException if {@code filter} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    Result filter(Predicate<Segment> filter);
+
+    /**
+     * Returns a mapping of the {@link Segment result segments} that are the results of a query against a database.  May be empty if the query did not return any segments.  A {@link Segment} can be
+     * only considered valid within a {@link Function mapping function} callback.
+     * <p>Consuming result {@link Segment segments} does not emit {@link Subscriber#onError(Throwable) error signals} from {@link Message#severity() error segments}.  Translation of error segments
+     * is subject to the {@code mappingFunction}.
+     *
+     * @param mappingFunction the {@link Function} that maps a {@link Segment} a to a value
+     * @param <T>             the type of the mapped value
+     * @return a mapping of the segments that are the results of a query against a database
+     * @throws IllegalArgumentException if {@code mappingFunction} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    <T> Publisher<T> flatMap(Function<Segment, ? extends Publisher<T>> mappingFunction);
+
+    /**
+     * Marker interface for a result segment.  Result segments represent the individual parts of a result from a query against a database.  Segments are typically update counts, data rows/{@code
+     * OUT} parameters, and messages of different severities.  Segments can also represent vendor-specific extensions.
+     *
+     * @since XXX
+     */
+    interface Segment {
+
+    }
+
+    /**
+     * Data segment consisting of {@link Row gettable data} and its {@link #metadata() RowMetadata}.  This segment can be backed by either a {@link Row} or any other type of gettable data such as
+     * {@code OUT} parameters.
+     *
+     * @since XXX
+     */
+    interface Data extends Segment, Row {
+
+        RowMetadata metadata();
+
+    }
```

Comment:
As commented here also: https://github.com/r2dbc/r2dbc-mssql/issues/199#issuecomment-823245179

While OUT parameters are a `Row`-ish data structure with `RowMetadata`-ish meta data, I think they deserve their own separate `Segment` subtype that may share some API with `Row` but allows for clearly distinguishing the two things for (at least) these two reasons:

- In the presence of both OUT parameters and result sets (see example below), it would otherwise not be easy to know whether the `Segment` corresponds to the result set(s) or the OUT parameters. This is specifically important in SQL Server, where there can be 0-N result sets, and finally some OUT parameters.
- While the `Row` and the `Row`-ish OUT parameters look similar, they might not be the same thing in ways that we cannot foresee now. Having 2 distinct types before a 1.0 release helps prevent breaking changes in the future.

Example of such a SQL Server procedure returning a result set and OUT parameters:

```sql
CREATE PROCEDURE test_proc
    @TheName nvarchar(50),
    @Greeting nvarchar(255) OUTPUT
AS
    SET NOCOUNT ON;  
    SET @Greeting = CONCAT('Hello ', @TheName);
    SELECT 1;
```

Potential naming suggestions depend on how much incompatible refactoring is still possible. Without any such refactorings, how about simply `RowSegment` and `OutSegment` (implying a rename of `UpdateCount` to `UpdateCountSegment` and of `Message` to `MessageSegment`). Many other options are possible, too.

Reply:
Introducing `OutParameters` as an aggregator for multiple out parameters seems a good direction, also introducing a `OutParametersSegment`.

As result, extracting `get` methods into a `Gettable` interface (basically what `Row` as of now) and making `Row` and `OutParameters` subtypes of `Gettable` feels like an actual improvement.

We can follow the `…Segment` naming in general. As a thought, moving `Segment` to a top-level type and letting e.g. `Row` implement `Segment` feels a bit overengineered.

Reply:
> As a thought, moving `Segment` to a top-level type and letting e.g. `Row` implement `Segment` feels a bit overengineered.

Not sure about the top-level type. But how can a `Row` exist outside of it being used as a `Segment`?